File: attention_service/storage/voice_store.py
import json

import numpy as np
from sqlalchemy import text

from config import get_engine


def save_voiceprint(candidate_id: str, embedding: np.ndarray) -> None:
    engine = get_engine()
    embedding_json = json.dumps(embedding.tolist())
    with engine.begin() as conn:
        conn.execute(
            text(
                """
                INSERT INTO voiceprints (candidate_id, embedding, created_at, updated_at)
                VALUES (:candidate_id, CAST(:embedding AS jsonb), NOW(), NOW())
                ON CONFLICT (candidate_id)
                DO UPDATE SET embedding = EXCLUDED.embedding, updated_at = NOW()
                """
            ),
            {"candidate_id": candidate_id, "embedding": embedding_json},
        )


def load_voiceprint(candidate_id: str) -> np.ndarray | None:
    engine = get_engine()
    with engine.connect() as conn:
        row = conn.execute(
            text(
                "SELECT embedding FROM voiceprints WHERE candidate_id = :candidate_id"
            ),
            {"candidate_id": candidate_id},
        ).fetchone()
    if not row:
        return None
    values = row[0]
    if isinstance(values, str):
        values = json.loads(values)
    return np.array(values, dtype=np.float32)


def delete_voiceprint(candidate_id: str) -> bool:
    engine = get_engine()
    with engine.begin() as conn:
        result = conn.execute(
            text("DELETE FROM voiceprints WHERE candidate_id = :candidate_id"),
            {"candidate_id": candidate_id},
        )
    return result.rowcount > 0

File: attention-node/src/services/voiceStore.js
import { prisma } from '../db/prisma.js'

export async function saveVoiceprint(candidateId, embedding) {
  await prisma.voiceprint.upsert({
    where: { candidateId },
    create: {
      candidateId,
      embedding,
    },
    update: {
      embedding,
      updatedAt: new Date(),
    },
  })
}

export async function loadVoiceprint(candidateId) {
  const row = await prisma.voiceprint.findUnique({ where: { candidateId } })
  if (!row) return null
  return row.embedding
}

export async function deleteVoiceprint(candidateId) {
  try {
    await prisma.voiceprint.delete({ where: { candidateId } })
    return true
  } catch {
    return false
  }
}

export function cosineSimilarity(a, b) {
  if (!Array.isArray(a) || !Array.isArray(b) || a.length !== b.length) {
    throw new Error('Invalid embeddings for comparison')
  }
  let sum = 0
  for (let i = 0; i < a.length; i += 1) sum += Number(a[i]) * Number(b[i])
  return sum
}
